src/app/OperationalSessionSetup.h

```diff
@@ -155,6 +155,34 @@ typedef void (*OnDeviceConnectionRetry)(void * context, const ScopedNodeId & pee
 class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate, public AddressResolve::NodeListener
 {
 public:
+    enum class State : uint8_t
+    {
+        Uninitialized,    // Error state: OperationalSessionSetup is useless
+        NeedsAddress,     // No address known, lookup not started yet.
+        ResolvingAddress, // Address lookup in progress.
+        HasAddress,       // Have an address, CASE handshake not started yet.
+        Connecting,       // CASE handshake in progress.
+        SecureConnected,  // CASE session established.
+        WaitingForRetry,  // No address known, but a retry is pending.  Added at
+                          // end to make logs easier to understand.
+    };
+
+    struct ConnnectionFailureInfo
+    {
+        const ScopedNodeId & peerId;
+        CHIP_ERROR error;
+        OperationalSessionSetup::State sessionState;
+
+        ConnnectionFailureInfo(const ScopedNodeId & peer, CHIP_ERROR err, OperationalSessionSetup::State state) :
+            peerId(peer), error(err), sessionState(state)
+        {}
+    };
+
+    using OnSetupSuccessful      = OnDeviceConnected;
+    using OnSetupFailure         = OnDeviceConnectionFailure;
+    using OnSetupRetry           = OnDeviceConnectionRetry;
+    using OnExtendedSetupFailure = void (*)(void * context, const ConnnectionFailureInfo & failureInfo);
```

Comment:
I would suggest naming the new thing OnSetupFailure and naming the old thing with the more awkward name.  In either case, "OnExtendedSetupFailure" sounds like "extended setup", whatever that is, failed.

Reply:
Use OnSetupFailure only for the new callback, providing aliases for the old API with awkward names is unnecessary;  removed that portion.